examples/getting_started/tools_and_agents.ipynb

```diff
@@ -0,0 +1,519 @@
+     "start_time": "2024-09-10T05:49:51.510653Z"
+    }
+   },
+   "outputs": [],
+   "source": [
+    "from duckduckgo_search import DDGS\n",
+    "\n",
+    "\n",
+    "def web_search(queries: list[str], max_results_per_query: int = 2) -> str:\n",
+    "    \"\"\"Performs web searches for given queries and returns URLs.\n",
+    "\n",
+    "    Args:\n",
+    "        queries: List of search queries.\n",
+    "        max_results_per_query: Maximum number of results to return per query.\n",
+    "\n",
+    "    Returns:\n",
+    "        str: Newline-separated URLs from search results or error messages.\n",
+    "\n",
+    "    Raises:\n",
+    "        Exception: If web search fails entirely.\n",
+    "    \"\"\"\n",
+    "    try:\n",
+    "        urls = []\n",
+    "        for query in queries:\n",
+    "            results = DDGS(proxies=None).text(query, max_results=max_results_per_query)\n",
+    "\n",
+    "            for result in results:\n",
+    "                link = result[\"href\"]\n",
+    "                try:\n",
+    "                    urls.append(link)\n",
+    "                except Exception as e:\n",
+    "                    urls.append(f\"{type(e)}: Failed to parse content from URL {link}\")\n",
+    "        return \"\\n\\n\".join(urls)\n",
+    "\n",
+    "    except Exception as e:\n",
+    "        return f\"{type(e)}: Failed to search the web for text\""
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "This `web_search` function is a more complex Tool that takes a list of search queries and returns a string of newline-separated URLs from the search results. It uses the DuckDuckGo search API to perform the searches."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## 5. Introduction to Agents\n",
+    "\n",
+    "Agents in Mirascope are autonomous or semi-autonomous entities that use LLMs and Tools to perform complex tasks or interact with users. They typically have:\n",
+    "\n",
+    "1. A state (e.g., conversation history, search history)\n",
+    "2. Access to one or more Tools\n",
+    "3. A method for interacting with an LLM\n",
+    "4. A main loop or execution flow\n",
+    "\n",
+    "Let's start implementing our WebSearchAgent:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 1,
+   "metadata": {
+    "ExecuteTime": {
+     "end_time": "2024-09-10T05:49:55.059472Z",
+     "start_time": "2024-09-10T05:49:54.640424Z"
+    }
+   },
+   "outputs": [],
+   "source": [
+    "from pydantic import BaseModel\n",
+    "\n",
+    "from mirascope.core import BaseMessageParam, openai\n",
+    "\n",
+    "\n",
+    "class WebAssistantBase(BaseModel):\n",
+    "    messages: list[BaseMessageParam | openai.OpenAIMessageParam] = []\n",
+    "    search_history: list[str] = []\n",
+    "    max_results_per_query: int = 2\n",
+    "\n",
+    "    def _web_search(self, queries: list[str]) -> str:\n",
```

Comment:
this will result in worse tool usage because the tool is generated from the function definition, so the docstring for `_web_search` is used here instead of the one for `web_search` (which is far more verbose).

Instead, we should implement the `web_search` tool to accept the `self` argument in preparation for including it in out agent.